### ISSUE TITLE
how to make copy networks in equinox

### ISSUE THREAD

**Issue by ak24watch** [patrick-kidger/equinox#1136]
_how to make copy of networks in equinox_, **like in dqn for target networks** please give example of simple network how it can be done

**Comment by johannahaffner** [comment]
An Equinox model is just a PyTree, you can make a copy using standard tools, e.g. with copy or `deepcopy`. You can also use `jtu.tree_map` with any update function, e.g. as in this snippet where the update is computed with standard gradient descent

```python
new_model = jax.tree_util.tree_map(lambda m, g: m - learning_rate * g, model, grads)
```

Does that clarify things?

**Comment by ak24watch** [comment]
> An Equinox model is just a PyTree, you can make a copy using standard tools, e.g. with copy or `deepcopy`. You can also use `jtu.tree_map` with any update function, e.g. as in this snippet where the update is computed with standard gradient descent
> 
> ```python
> new_model = jax.tree_util.tree_map(lambda m, g: m - learning_rate * g, model, grads)
> ```
> 
> Does that clarify things?

**but equinox model also as static things function info ?
 I dont update just need to copy first  then I should just keep it as it is in treemap , does it copy ? ** 

like 
```
model_params, static = eqx.partiton(model) 
new_model_params = jax.tree_util.tree_map(lambda m: m, model_params)
new_model = eqx.combine(new_model_params, static)
```

**Comment by johannahaffner** [comment]
Yes, this will create a new PyTree with the same values as the old one.

**Comment by lockwo** [comment]
In the double dqn case, it would look something like 


<details>

<summary> inefficient code </summary>

```python
import time
from dataclasses import dataclass
import equinox as eqx
import jax
import jax.numpy as jnp
import numpy as np
import optax
import matplotlib.pyplot as plt
import gymnasium as gym



class QNet(eqx.Module):
    mlp: eqx.nn.MLP

    def __init__(self, obs_dim: int, n_actions: int, width: int, depth: int, key):
        self.mlp = eqx.nn.MLP(
            in_size=obs_dim,
            out_size=n_actions,
            width_size=width,
            depth=depth,
            key=key,
        )

    @eqx.filter_jit
    def __call__(self, x: jax.Array) -> jax.Array:
        return self.mlp(x)

@dataclass
class ReplayBuffer:
    capacity: int
    obs_dim: int
    ptr: int = 0
    size: int = 0

    def __post_init__(self):
        self.obs = np.zeros((self.capacity, self.obs_dim), dtype=np.float32)
        self.next_obs = np.zeros((self.capacity, self.obs_dim), dtype=np.float32)
        self.acts = np.zeros((self.capacity,), dtype=np.int32)
        self.rews = np.zeros((self.capacity,), dtype=np.float32)
        self.dones = np.zeros((self.capacity,), dtype=np.float32)

    def add(self, s, a, r, s2, done):
        i = self.ptr
        self.obs[i] = s
        self.acts[i] = a
        self.rews[i] = r
        self.next_obs[i] = s2
        self.dones[i] = float(done)
        self.ptr = (i + 1) % self.capacity
        self.size = min(self.size + 1, self.capacity)

    def sample(self, batch_size: int):
        idx = np.random.randint(0, self.size, size=batch_size)
        batch = {
            "s": self.obs[idx],
            "a": self.acts[idx],
            "r": self.rews[idx],
            "s2": self.next_obs[idx],
            "not_done": 1.0 - self.dones[idx],
        }
        return {k: jnp.asarray(v) for k, v in batch.items()}


@eqx.filter_jit
def train_step(online: QNet, target: QNet, opt_state, batch):

    def loss_fn(model: QNet):
        q_s = eqx.filter_vmap(model)(batch["s"])
        q_sa = jnp.take_along_axis(q_s, batch["a"][:, None], axis=-1).squeeze()

        q_s2_online = eqx.filter_vmap(model)(batch["s2"])
        next_actions = jnp.argmax(q_s2_online, axis=1)

        q_s2_tgt_all = eqx.filter_vmap(target)(batch["s2"])
        q_s2_tgt = jnp.take_along_axis(q_s2_tgt_all, next_actions[:, None], axis=-1).squeeze()

        y = batch["r"] + GAMMA * batch["not_done"] * q_s2_tgt
        td = q_sa - y
        return 0.5 * jnp.mean(td * td)

    loss, grads = eqx.filter_value_and_grad(loss_fn)(online)
    updates, opt_state = optimizer.update(
        grads, opt_state, eqx.filter(online, eqx.is_array)
    )
    online = eqx.apply_updates(online, updates)
    return online, opt_state, loss



@eqx.filter_jit
def soft_update(target: QNet, online: QNet, tau: float) -> QNet:
    tar = eqx.filter(target, eqx.is_array)
    on, off = eqx.partition(online, eqx.is_array)
    new_arr = jax.tree.map(lambda t, o: t * (1.0 - tau) + o * tau, tar, on)
    return eqx.combine(new_arr, off)


class LinearSchedule:
    def __init__(self, eps_start: float, eps_end: float, decay_steps: int):
        self.eps_start = eps_start
        self.eps_end = eps_end
        self.decay_steps = max(1, decay_steps)

    def __call__(self, t: int) -> float:
        frac = min(1.0, t / self.decay_steps)
        return self.eps_start + frac * (self.eps_end - self.eps_start)



SEED = 0
TOTAL_STEPS = 20_000
START_TRAINING_AFTER = 1_000
BATCH_SIZE = 512
GAMMA = 0.99
LR = 3e-4
WIDTH = 128
DEPTH = 2
BUFFER_CAP = 10_000
TAU = 0.005

key = jax.random.PRNGKey(SEED)
np.random.seed(SEED)

env = gym.make("CartPole-v1")
obs, _ = env.reset(seed=SEED)
obs = np.asarray(obs, dtype=np.float32)

obs_dim = int(obs.shape[-1])
n_actions = int(env.action_space.n)

key, k1, _ = jax.random.split(key, 3)
online = QNet(obs_dim, n_actions, WIDTH, DEPTH, k1)
target = QNet(obs_dim, n_actions, WIDTH, DEPTH, k1)
optimizer = optax.adam(LR)
opt_state = optimizer.init(eqx.filter(online, eqx.is_array))

buf = ReplayBuffer(BUFFER_CAP, obs_dim)
eps = LinearSchedule(1.0, 0.05, decay_steps=TOTAL_STEPS // 2)

ep_ret, returns = 0.0, []
t0 = time.time()

for t in range(1, TOTAL_STEPS + 1):
    if t % (TOTAL_STEPS // 25) == 0:
        print(t)
    if np.random.rand() < eps(t):
        act = env.action_space.sample()
    else:
        q = online(jnp.asarray(obs))
        act = int(jnp.argmax(q))

    step_out = env.step(act)
    if len(step_out) == 5:
        next_obs, rew, terminated, truncated, _ = step_out
        done = bool(terminated or truncated)
    else:
        next_obs, rew, done, _ = step_out
    next_obs = np.asarray(next_obs, dtype=np.float32)

    buf.add(obs, act, float(rew), next_obs, done)
    obs = next_obs
    ep_ret += float(rew)

    if buf.size >= START_TRAINING_AFTER:
        batch = buf.sample(BATCH_SIZE)
        online, opt_state, _ = train_step(online, target, opt_state, batch)

    target = soft_update(target, online, TAU)

    if done:
        obs, _ = env.reset()
        obs = np.asarray(obs, dtype=np.float32)
        returns.append(ep_ret)
        ep_ret = 0.0

plt.figure()
plt.plot(returns, label="Return per episode")
plt.xlabel("Episode")
plt.ylabel("Return")
plt.legend()
plt.tight_layout()
plt.show()
```

<\details>

**Comment by ak24watch** [comment]
> In the double dqn case, it would look something like
> 
> inefficient code

yes i got it , thank you

**Comment by lockwo** [comment]
That's quite a long piece of code. You might have more luck getting assistance with a focused MVC. Glancing at it, if you are running on GPU, lots of .at[].set can be slow on GPU (indexing memory can sometimes be slower than doing even redundant calculations via efficient kernels), but idk.

**Comment by ak24watch** [comment]
> In the double dqn case, it would look something like
> 
> inefficient code

how would you calculate episode return when you have vectorized envs ?
